### PR TITLE
Flysystem support added.

### PR DESCRIPTION
**Pull request by graundas**
Hi. I have added support for Flysystem filesystem support. I have created a Flysystem loader and some tests. For more details check docs I updated.

### CODE REVIEW

File: Binary/Loader/FlysystemLoader.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    public function __construct(
+        ContainerInterface $container,
+        ExtensionGuesserInterface $extensionGuesser,
+        $sFileSystem)
+    {
+        $this->extensionGuesser = $extensionGuesser;
+        
+        try {
+            $sFileSystemService = sprintf('oneup_flysystem.%s_filesystem', $sFileSystem);
+            $this->filesystem = $container->get($sFileSystemService);
```
Comment (makasim) [inline Binary/Loader/FlysystemLoader.php]: Can we inject this service directly. Injecting of the container and 
"guessing" the filesystem service in the constructor looks strange to me.

File: Binary/Loader/FlysystemLoader.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    public function __construct(
+        ContainerInterface $container,
+        ExtensionGuesserInterface $extensionGuesser,
+        $sFileSystem)
+    {
+        $this->extensionGuesser = $extensionGuesser;
+        
+        try {
+            $sFileSystemService = sprintf('oneup_flysystem.%s_filesystem', $sFileSystem);
```
Comment (makasim) [inline Binary/Loader/FlysystemLoader.php]: `oneup_flysystem.%s_filesystem` - looks like it is a service from a bundle. What if someone wants to use this loader with manually created filesystem service?

File: DependencyInjection/Factory/Loader/FlysystemLoaderFactory.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+class FlysystemLoaderFactory implements LoaderFactoryInterface
+{
+    /**
+     * {@inheritDoc}
+     */
+    public function create(ContainerBuilder $container, $loaderName, array $config)
+    {
+        $loaderDefinition = new DefinitionDecorator('liip_imagine.binary.loader.prototype.flysystem');
+        $loaderDefinition->replaceArgument(2, $config['file_system']);
```
Comment (makasim) [inline DependencyInjection/Factory/Loader/FlysystemLoaderFactory.php]: inject service here

**Comment by Sander-Toonen** [comment]
Hi, I was just thinking to pick development on [this issue](https://github.com/1up-lab/OneupFlysystemBundle/issues/56) up. What is your point of view regarding stream support in OneupFlysystemBundle and what would be the up/downside of using this approach?

**Comment by makasim** [comment]
I am fine with Flysystem loader if it does not coupled to the bundle. See my comments.

**Comment by graundas** [comment]
I agree with you, @makasim . I will rewrite my code in a few days, so that a service would be injected instead of getting It from the container. @Sander-Toonen , I don't think a wrapper would be a better solution, because we would have to register a global php wrapper, which, in my opinion, is a slower and less configurable solution compared to dependency injection.

**Comment by graundas** [comment]
Hi. I have updated this Loader to use a service, which is being injected now. Check docs.
Travis-CI check has failed, because Flysystem library requires PHP 5.4.0 and Travis-CI checks compatibility with PHP 5.3. I am not sure if I should update .travis.yml by removing PHP 5.3 check. Should I do that?

**Comment by makasim** [comment]
Check php version and do composer require flysystem only if php version is higher than 5.3.

**Comment by makasim** [comment]
also add skip method to the tests, if php version is 5.3

**Comment by makasim** [comment]
Let's discuss drop of php5.3 support here: https://github.com/liip/LiipImagineBundle/issues/679

File: Resources/config/imagine.xml [hunk 4]
```diff
@@ -205,6 +206,11 @@
             <argument><!-- will be injected by StreamLoaderFactory --></argument>
             <argument><!-- will be injected by StreamLoaderFactory --></argument>
         </service>
+        
+        <service id="liip_imagine.binary.loader.prototype.flysystem" class="%liip_imagine.binary.loader.flysystem.class%">
```
Comment (makasim) [inline Resources/config/imagine.xml]: add abstract=true

File: composer.json [hunk 5]
```diff
@@ -38,7 +38,8 @@
         "aws/aws-sdk-php": "~2.4",
         "amazonwebservices/aws-sdk-for-php": "~1.0",
         "psr/log": "~1.0",
-        "ext-gd": "*"
+        "ext-gd": "*",
+        "league/flysystem": "~1.0"
     },
 
     "suggest": {
```
Comment (makasim) [inline composer.json]: add it to suggest too.

**Comment by graundas** [comment]
All you asked is done. Will you merge It to your repository?

File: Binary/Loader/FlysystemLoader.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    protected $filesystem;
+
+    /**
+     * @var ExtensionGuesserInterface
+     */
+    protected $extensionGuesser;
+
+    public function __construct(
+        ExtensionGuesserInterface $extensionGuesser,
+        Filesystem $filesystem)
```
Comment (makasim) [inline Binary/Loader/FlysystemLoader.php]: Is there interface?

Comment (makasim) [inline Binary/Loader/FlysystemLoader.php]: ) {

Comment (graundas) [inline Binary/Loader/FlysystemLoader.php]: Are you talking about "Filesystem $filesystem" on line 24?

File: Resources/doc/data-loader/flysystem.rst [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+                local:
+                    directory:  "path/to/profile/photos"
+
+        filesystems:
+            profile_photos:
+                adapter: profile_photos
+
+
+.. _`Flysystem`: https://github.com/thephpleague/flysystem
+.. _`OneupFlysystemBundle`: https://github.com/1up-lab/OneupFlysystemBundle
```
Comment (makasim) [inline Resources/doc/data-loader/flysystem.rst]: @javiereguiluz does the doc looks ok?

File: Tests/Binary/Loader/FlysystemLoaderTest.php [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+            $loader->find('assets/cats.jpeg')->getContent()
+        );
+    }
+
+    /**
+     * @depends testCouldBeConstructedWithExpectedArguments
+     */
+    public function testThrowsIfInvalidPathGivenOnFind($loader)
+    {
+        $sPath = 'invalid.jpeg';
```
Comment (makasim) [inline Tests/Binary/Loader/FlysystemLoaderTest.php]: with `s` prefix. just `$path`.

File: composer.json [hunk 9]
```diff
@@ -45,7 +45,8 @@
         "twig/twig":                "If you'd want to use some handy twig filters, version 1.12 or greater required",
         "aws/aws-sdk-php":          "Add it if you'd like to use aws v2 or v3 resolver",
         "amazonwebservices/aws-sdk-for-php": "Add it if you'd like to use aws v1 resolver",
-        "monolog/monolog":          "If you'd want to write logs"
+        "monolog/monolog":          "If you'd want to write logs",
+        "league/flysystem":         "If you'd want to test Flysystem support"
```
Comment (makasim) [inline composer.json]: If you'd want to use Flysystem binary loader. or something like that

**Comment by graundas** [comment]
Hi. All done.

**Comment by Sander-Toonen** [comment]
Very nice work @graundas. Can't wait to see this merged.

**Comment by makasim** [comment]
Looks good to me. @javiereguiluz Could you please have a look at the docs?

**Comment by heikokrebs** [comment]
:+1:

**Comment by makasim** [comment]
@javiereguiluz ping, @graundas could you please rebase the PR?

File: Resources/doc/data-loader/flysystem.rst [hunk 10]
```diff
@@ -0,0 +1,35 @@
+FlysystemLoader
+============
```
Comment (javiereguiluz) [inline Resources/doc/data-loader/flysystem.rst]: This line should be as long as the preceding heading.

File: Resources/doc/data-loader/flysystem.rst [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+This loader lets you load images from `Flysystem`_ filesystem abstraction layer,
+which can be used in Symfony projects by installing, for example, `OneupFlysystemBundle`_.
+
+Value of 'filesystem_service' property must be a service,
```
Comment (javiereguiluz) [inline Resources/doc/data-loader/flysystem.rst]: replace this:

```
'filesystem_service'
```

by this (backticks must be doubled):

```
``filesystem_service``
```

File: Resources/doc/data-loader/flysystem.rst (hunk 7 above)
Comment (javiereguiluz) [inline Resources/doc/data-loader/flysystem.rst]: I've added some minor comments, but most of it is correct.

**Comment by graundas** [comment]
Hi. I have just rebased my PR. Docs corrected also.

**Comment by makasim** [comment]
Thanks @graundas for the contribution and @javiereguiluz  for the review.

**Pull request merged**